### DISCUSSION TITLE
Self-paced reading - button plugin

### DISCUSSION THREAD

**Discussion by gustavoestivalet** [jspsych/jsPsych#895] [Q&A] [answered]
Hello Guys,

I am starting a new topic to discuss a **button plugin** for the **self-paced reading** experiment.

The video from MRO was very clear for the development of a moving window plugin with the keyboard presenting word-by-word. Than, the discussion in `https://github.com/jspsych/jsPsych/discussions/857` allows to transform it in a more general word/phrase moving window experiment with many stimuli. It worked perfectly!

Now, I am thinking in the application of mobile experiments, therefore, it would be great to transform the "keyboard-moving-window" plugin into a **"button-moving-window"** plugin. I took a look in the "html-button-response" plugin to use it, but the structure is very different from the "html-keyboard-response" used in the "moving-window" plugin, specially regarding a) the button display and b) RT measure.

Therefore, I wonder if you could help to develop the **"button-moving-window"** plugin. In general, in the trial event with this plugin, we have to a) present a button "Next" which will move the window and b) catch all the RT from this button in an array.

Thank you so much for all you assistance and support,
Best,

**Comment by jodeleeuw** [comment, marked as answer]
Hi @gustavoestivalet ,

You need to do two things: (1) add the button to the screen, (2) execute a function when the button is clicked.

Adding the button to the screen is a little tricky because you need the button to persist on the screen while the rest of the stimulus is changing. I'd suggest something like this at the start of the plugin

```js
display_element.innerHTML = "<div id='moving-window-stimulus'></div>" +
	"<button id='moving-window-next' class='jspsych-btn'>Next</button>"
```

This sets up a `<div>` element that we can modify without erasing the button.

When you update the stimulus on the screen, instead of `display_element.innerHTML = ...` you'll use `display_element.querySelector('#moving-window-stimulus').innerHTML = ...` to change the contents of this `<div>`. 

To set up the button to control the trial:

```js
display_element.querySelector('#moving-window-next').addEventListener('click', after_response);
```

This will call the `after_response` function whenever the button is clicked. You will need to modify that function from the keyboard version to measure response time. You can have a variable that keeps track of the last time the button was pressed and update it whenever the button is pressed. You can subtract the current time from the last time to get the RT.

```js
var last_button_press = performance.now();
function after_response(){
	var rt = performance.now() - last_button_press;
	last_button_press = performance.now();
	// ... other code here ...
}
```

**Reply by gustavoestivalet** [reply]
Hi @jodeleeuw , thank you so much for your advice. I started a new plugin from scratch with your hints, and it seems it is really doing the job! I am just not sure if the RT calculation is correct, could you please check it out? Also, It seems to me that the last block "add rt" could be simplified and optimized, could you take a look please?

PLUGIN: moving-window-button-phrase
```
jsPsych.plugins["moving-window-button-phrase"] = (function() {

  var plugin = {};

  plugin.info = {
    name: 'moving-window-button-phrase',
    description: '',
    parameters: {
      words: {
        type: jsPsych.plugins.parameterType.STRING,
        default: undefined
      },
    }
  }

  plugin.trial = function(display_element, trial) {

    // data saving
    var trial_data = {
		words: trial.words
    };
	var rt = [];
	var current_word = 0;
	
	// create stimuli
	function create_moving_window(words, position){
		var word_list = words;
		var words = word_list.map(function(word, index){
			if(index == position){
				return word;
			} else {
				var output_string = "";
				for(var i = 0; i < word.length; i++){
					if(word[i] != " "){
						output_string += "-";
					} else {
						output_string += " ";
					}
				}
				return output_string;
			}
		}).join(" ");
		return words;
	}
	
	// display stimulus
	function show_words(){
		display_element.innerHTML = "<div id='moving-window-phrase'></div>" +
		"<button id='moving-window-next' class='jspsych-btn'>PRÓXIMO</button>";
		
		display_element.querySelector('#moving-window-phrase').innerHTML = "<p style='font-family: monospace;'>"+create_moving_window(trial.words, current_word)+"</p>";
		
		// add event listeners to buttons
		display_element.querySelector('#moving-window-next').addEventListener('click', after_response);
	}

	// add rt
	var last_button_press = performance.now();
	function after_response(){
		rt[current_word] = performance.now() - last_button_press;
		last_button_press = performance.now();
		current_word++;
		if(current_word == trial.words.length){
			end_trial();
		}else{
			show_words();
		}
	}
	
	// end trial
	function end_trial(){
		trial_data.rt = JSON.stringify(rt);

		jsPsych.finishTrial(trial_data);
	}

	show_words();

  };

  return plugin;
})();

```
**EXPERIMENT:**
```
<!DOCTYPE html>
<html>
	<head>
		<meta charset="utf-8">
		<title>Self-Paced Reading Move Window - Button</title>
		<script src="jspsych/jspsych.js"></script>
		<script src="jspsych/plugins/jspsych-html-button-response.js"></script>
		<script src="jspsych-moving-window-button-phrase.js"></script>
		<link href="jspsych/css/jspsych.css" rel="stylesheet" type="text/css">
	</head>
	<body></body>

	<script>
	// create timeline
	var timeline = [];

	// create practice_stimuli
	var practice_stimuli = [
		{words: ["O garoto","ganhou","o dinheiro.","A poupança","ajudou","nos gastos.","Ele","então comprou","o brinquedo."], question: "AS ASAS",data: {correct_response: "q",stimulus: "O garoto/ganhou/o dinheiro./A poupança/ajudou/nos gastos./ Ele/então comprou/o brinquedo.",question: "AS ASAS"}},
		{words: ["A menina","escutou","os tios.","A família","interveio","no caso.","Isso","mudou","a situação."], question: "OS TIOS",data: {correct_response: "p",stimulus: " A menina/escutou/os tios./A família/ interveio/no caso./Isso/mudou/a situação.",question: "OS TIOS"}}
	];

	// create practice
	var practice = {
		type: "moving-window-button-phrase",
		words: jsPsych.timelineVariable("words"),
		data: jsPsych.timelineVariable("data")
	};

	// create practice_question
	var practice_question = {
		type: "html-button-response",
		stimulus: jsPsych.timelineVariable("question"),
		choices: ["NÃO", "SIM"],
		data: jsPsych.timelineVariable("data"),
		post_trial_gap: 500
	};

	// create practice_procedure
	var practice_procedure = {
		timeline: [practice, practice_question],
		timeline_variables: practice_stimuli,
		randomize_order: true
	};
	// put practice_procedure in timeline
	timeline.push(practice_procedure);

	// run experiment
	jsPsych.init({
		timeline: timeline,
		on_finish: function(){
			jsPsych.data.displayData();
		}
	});

	</script>
</html>
```

**Reply by jodeleeuw** [reply]
This looks great to me. I don't think I'd change a thing! Is it giving you sensible results?

**Reply by gustavoestivalet** [reply]
Hi @jodeleeuw , that sounds really good! In my own pilot tests, it seems it is working good. I will probably have some pilot participants until Friday, than I will have a better picture of the RTs. I will keep you updated!

Quite changing the topic, as you probably saw, Brazilian Portuguese use diacritics and I am having some problems in the MySql database. I already put `<meta charset="utf-8">` in the head of the experiment but it only solved the display. I already set all MySql collation to "utf8mb4_unicode_ci" everywhere, but no solution. Is there anything in jsPsych that should be done to correctly render uft-8 diacritics? Otherwise, any further hint in how to solve it?

Thank a lot for everything!

**Reply by gustavoestivalet** [reply]
Hi, Guys, just to finish this topic (by the moment), I updated the PHP version in my server to 7.4 and all character/collation errors were corrected. Done!
